### PR TITLE
feat: prepareSpendCallData

### PR DESCRIPTION
**Pull request by fan-zhang-sv**
### _Summary_

# Add prepareSpendCallData method for spend permission execution

## Summary

This PR introduces a new utility method `prepareSpendCallData` that simplifies the process of preparing call data for executing spend permissions. This method generates the necessary transaction calls for both approving and spending from a spend permission in a single operation.

### 🆕 New Features

- **`prepareSpendCallData` method**: New utility function that prepares call data for spend permission execution
  - Automatically handles both `approveWithSignature` and `spend` operations
  - Supports custom spend amounts or uses remaining allowance by default
  - Returns properly formatted call data for use with `wallet_sendCalls`
  - Includes comprehensive error handling and validation

### 📝 API Changes

- **Export**: Added `prepareSpendCallData` to the public spend-permission API
- **Type Safety**: Introduced `PrepareSpendCallDataResponseType` for type-safe return values

### 🧪 Testing

- **Comprehensive test coverage**: 271 lines of tests covering:
  - Basic functionality with default and custom amounts
  - Error handling (zero amounts, permission errors, encoding errors)
  - Edge cases (large amounts, inactive permissions)
  - Proper integration with existing utilities (`getPermissionStatus`, `toSpendPermissionArgs`)
  - Mock validation for all dependencies

### Implementation

The `prepareSpendCallData` function:
1. Fetches current permission status to determine remaining allowance
2. Uses provided amount or defaults to remaining spend amount
3. Validates spend amount is greater than zero
4. Converts permission to contract-compatible format
5. Encodes two function calls:
   - `approveWithSignature`: Approves the permission (no-op if already approved)
   - `spend`: Executes the spend operation
6. Returns array of call objects ready for `wallet_sendCalls`

### Usage Example

```typescript
import { prepareSpendCallData } from '@base-org/account/spend-permission';

// Prepare calls for spending a specific amount
const calls = await prepareSpendCallData(permission, 50n * 10n ** 6n);

// Or spend all remaining allowance
const callsFullAmount = await prepareSpendCallData(permission);

// Execute with wallet_sendCalls
await provider.request({
  method: 'wallet_sendCalls',
  params: [{
    version: '2.0.0',
    atomicRequired: true,
    from: permission.permission.spender,
    chainId: `0x${permission.chainId?.toString(16)}`,
    calls,
  }],
});
```

## Benefits

- **Developer Experience**: Simplifies spend permission execution to a single method call
- **Safety**: Automatically includes approval step, ensuring permissions are properly activated
- **Flexibility**: Supports both custom amounts and automatic full-allowance spending
- **Type Safety**: Fully typed with comprehensive error handling
- **Atomic Operations**: Returns call data designed for atomic execution via `wallet_sendCalls`

## Related

This complements the existing spend permission utilities:
- `requestSpendPermission`: Creates new spend permissions
- `prepareRevokeCallData`: Revokes existing permissions (from `felix/revokes` branch)
- `fetchPermissions`: Retrieves existing permissions
- `getPermissionStatus`: Checks permission status and remaining allowance

#### succeeded spend tx

image

### CODE REVIEW

File: examples/testapp/src/pages/payment/index.page.tsx [hunk 1]
```diff
@@ -127,7 +127,8 @@ export default function Payment() {
       } else {
         toast({
           title: 'Payment failed',
-          description: 'error' in result ? result.error : 'Unknown error',
+          description:
+            'error' in result && typeof result.error === 'string' ? result.error : 'Unknown error',
```
Comment (fan-zhang-sv) [inline examples/testapp/src/pages/payment/index.page.tsx:131]: CI was broken, might be some disparity on CI and local, CI wasn't able to identify the discriminated union

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/prepareSpendCallData.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      to: spendPermissionManagerAddress,
+      data: approveData,
+      value: '0x0', // explicitly set to 0x0
+    },
+    {
+      to: spendPermissionManagerAddress,
+      data: spendData,
+      value: '0x0', // explicitly set to 0x0
+    },
+  ];
```
[the remark this answers is not in the document]
Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/prepareSpendCallData.ts:129]: makes sense, i think we should convey to the customers that both `eth_sendTransaction` and `wallet_sendCalls` can do the job here

i don't think we can do much at code level, but we can be more explicitly about this
* updated docString to provide examples for both methods
* updated the hackMD doc to be provide examples for both methods + what's the difference

lmk wdyt

[the remark this answers is not in the document]
Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/prepareSpendCallData.ts]: discussed offline w spencer, we are going for this commit [3c40ed4](https://github.com/base/account-sdk/pull/80/commits/3c40ed43b3117503b689d791c0bb79e3f960ed1b)

**Review by montycheese — APPROVED** [review]
